### PR TITLE
[WIP] Smagorinsky turbulence closure: first try

### PR DESCRIPTION
**Pull request by jm-c**
A preliminary implementation of (constant) smagorinsky.
- temporary arrays to be revisited
- need to improve the way to disable it (without loss of performance)
- no attempt to try it on GPU. 
Tested with deep_convection example, running nicely with smag_coeff=3.3e-2 and Prandtl_num=1

### CODE REVIEW

**Comment by glwagner** [comment]
Awesome!

We can use multiple dispatch coupled with a new type that represents a “TurbulenceModel” to achieve the ability to disable / enable LES. 

I’ll demonstrate

**Comment by glwagner** [comment]
Awesome!

We can use multiple dispatch coupled with a new type that represents a “TurbulenceModel” to achieve the ability to disable / enable LES. 

I’ll demonstrate what I mean.

**Comment by glwagner** [comment]
@jm-c can you take a look at the pull request for a new time stepper type?

If you approve those changes, I will merge them into this pull request.

File: src/fields.jl [hunk 1]
```diff
@@ -108,15 +65,26 @@ struct EdgeField{A<:AbstractArray, G<:Grid} <: Field
...
-    data = zeros(T, size(g))
-    EdgeField{typeof(data), typeof(g)}(data, g)
+EdgeField(T, arch, grid) = EdgeField(zeros(T, arch, grid), grid)
+EdgeField(arch, grid) = EdgeField(zeros(arch, grid), grid)
+
+abstract type Location end
+struct Cell <: Location end
+struct Interface <: Location end
+
+"""
```
Comment (glwagner) [inline src/fields.jl]: @jm-c See the new `GeneralField` type here.

File: src/turbulence_closures.jl [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+S₁₂ :: S₂₁ :: FFC
+S₁₃ :: S₃₁ :: FCF
+S₂₃ :: S₃₂ :: CFF
+
+finally,
+
+S66 = S₁₁² + S₂₂² + S₃₃².
+=#
+
+struct TurbulentViscosity{A, G}
```
Comment (glwagner) [inline src/turbulence_closures.jl]: @jm-c see my new proposal notation here. Feel free to change.

File: src/fields.jl (hunk 1 above)
Comment (glwagner) [reply src/fields.jl]: Note that once this is done we should also be able to eliminate `CellField` and `FaceFieldX`, etc, replacing them with `GeneralField` (and possibly renaming to just "`Field`"). 

This should be easy via the definition of 

`const CellField = GeneralField{C, C, C}`, 

with no other code changes, I think. Which is nice.

**Comment by johncmarshall54** [comment]
Greg, Distinction between edgefield, cellfield etc, was a source of
confusion for Ali as oceananigans was built. Concerned about abstracting to
such a level that we can't easily see what's going on. We should all meet
in person to talk about how we reinstate finite volumes and support more
general grids. We can talk about nomenclature at the same time. When we get
back from Caltech? John

**Comment by glwagner** [comment]
Happy to discuss further.

The `GeneralField` is a *solution* to the problem / confusion you are referencing. As such, its intent is to reduce, rather than create, confusion. @jm-c and I (and some others) discussed this solution some time ago -- I thought I created an issue about it then, but I can't find it now -- but it was never implemented.

**Comment by ali-ramadhan** [comment]
Closing this PR as a simpler/lighter implementation of constant Smagorinsky has been included in v0.9.0.

**Pull request closed without merging**